### DISCUSSION TITLE
Changing turbine types for few turbines in wind farm

### DISCUSSION THREAD

**Discussion by Teddybear0227** [NatLabRockies/floris#766] [Q&A] [answered]
I am struggling to find a way to change the turbine type for few turbines in a wind farm.

My project aims to calculate the AEP and power vs wind direction, analysing the interaction between three nearby wind farms. 

My simulation considers 215 wind turbines in total, and wind turbines from 186-215 in layout_x and layout_y use different type of wind turbine to wind turbines from 1-185. 

In the current state, my simulation only considers one type of wind turbine (Siemens SWT-3.6MW-107) but I would like to assign another wind turbine type to wind turbines from 186-215 in layout_x and layout_y.

**Comment by misi9170** [comment]
Hi @Teddybear0227, thanks for your question. I'd first recommend taking a look at example 17_mutliple_turbine_types.py. You'll see that this example uses the floris input yaml gch_multiple_turbine_types.yaml (in examples/inputs). In the that yaml, in the `turbine_types` field of the `farm` field, you'll see that two turbine models are listed: `nrel_5MW` and `iea_10MW`. 

The `turbine_types` field can either contain a single turbine model, which is the "usual" case, in which case all turbines are assumed to be the same model; or it can contain a list of N turbines, where N is the number of turbine locations (so the list in `turbine_types` should have the same length as the list in `layout_x` and `layout_y`). The latter is the case in gch_multiple_turbine_types.yaml, and is likely what you'll want to do. So, in your case, you'd want to put into `turbine_types` 185 copies of the first turbine type and then 29 of the other turbine type, with locations specified in `layout_x` and `layout_y`. 

The relevant section of your FLORIS input yaml will look something like (with made-up x and y locations):
```
farm:
  layout_x:
  - 0.0
  - 630.0
  - 1260.0
  .
  .
  .
  - 5000.0
  - 10000.0
  .
  .
  .
  - 10630.0
  layout_y:
  - 0.0
  - 0.0
  - 0.0
  .
  .
  .
  - 0.0
  - 1000.0
  .
  .
  .
  - 1000.0
  turbine_type:
  - turbine_type_1
  - turbine_type_1
  - turbine_type_1
  . 
  .
  . 
  - turbine_type_1
  - turbine_type_2
  .
  .
  . 
  - turbine_type_2
```

By default, FLORIS will look for those turbine types in the floris/simulation/turbine_library/ directory. So, you could simply place the turbine model yamls in the turbine_library directory. Alternatively, you can use the `!include` notation to point to another directory where your turbine models are stored (for an example, see gch_fixed.yaml in examples/inputs_floating/), in which case your FLORIS input yaml will look more like
```
farm:
  layout_x:
  - 0.0
  - 630.0
  - 1260.0
  .
  .
  .
  - 5000.0
  - 10000.0
  .
  .
  .
  - 10630.0
  layout_y:
  - 0.0
  - 0.0
  - 0.0
  .
  .
  .
  - 0.0
  - 1000.0
  .
  .
  .
  - 1000.0
  turbine_type:
  - !include path/to/turbine_type_1
  - !include path/to/turbine_type_1
  - !include path/to/turbine_type_1
  . 
  .
  . 
  - !include path/to/turbine_type_1
  - !include path/to/turbine_type_2
  .
  .
  . 
  - !include path/to/turbine_type_2
```

**Reply by misi9170** [reply, marked as answer]
Also, note that for this to work most smoothly, I'd recommend including the full layout of the farm in the floris input yaml file (in the `layout_x` and `layout_y` fields), rather than setting them in `fi.reinitialize()`. Otherwise, you may receive an error that the length of the `turbine_types` field does not match the length of `layout_x` when you initially load the model.

**Reply by Teddybear0227** [reply]
Thank you for the detailed reply @misi9170 . I have tried the second method and it worked perfectly!! Thank you very much. I have also discovered that I can do this via reinitialise using the following code.

turbine_types = ["SWT_3.6.yaml"] * 185 + ["V80.yaml"] * (214 - 185 + 1)

# Now, redefine and add the custom turbine types
fi.reinitialize(
    turbine_type=turbine_types
)

Thank you again for your support!!
